lib/Driver/ToolChain.cpp

```diff
@@ -280,6 +280,32 @@ mergeBatchInputs(ArrayRef<const Job *> jobs,
   return false;
 }
 
+/// Explicitly (re)order \p jobs into \p sortedJobs, ensuring they are a
+/// _subsequence_ of the sequence of input files in \p C (returned by \c
+/// Compilation::getInputFiles), so that loops that select primaries by an
+/// "is-primary" predicate on the top-level file list, and those that directly
+/// iterate over the elements in a batch job's output list, always get the same
+/// sequence.
+static void
+sortBatchInputJobs(ArrayRef<const Job *> jobs,
+                   SmallVector<const Job *, 16> &sortedJobs,
+                   Compilation &C) {
+  llvm::StringMap<const Job *> jobsByInput;
```

Comment:
Mapping by string seems expensive. Is it not possible to just use the order of jobs before shuffling?

Reply:
No, but I can perform a shuffle in a way that maintains the order if you prefer. I would like to leave this in as an assert that the orders match, in that case.

Reply:
Maybe it doesn't matter cause we do tons of other string lookups. This just wasn't the obvious way to get the original order back for me.